File: Matlab Version (Deprecated)/App and Required Functions/ECPC_V1.m
% EECPC - extrusion mode of CCC by replacing mobile phase with SP
% written by Hoon Choi

% Assume the same flowrate is used
% Turnover time is automatically calculated 
% extrution state: turnover = Ncup, dT = vcup
% Xtot: total MP profiles (CM+EECPC)
% Ytot: total SP profiles (CM+EECPC)
% Cout: elution volume vs. column outlet concentration
% X: MP profiles during EECPC
% Y: SP profiles during EECPC
% Xcm: MP profiles from CM
% Ycm: SP profiles from CM


function [Vspan, Cout, Xtot, Ytot, V_boundary] = ECPC_V1(KD, Vc, Sf, Xcm, Ycm)

P = Sf/(1-Sf); %phase ratio
Vm = Vc*(1-Sf); % MP volume
% Vs = Vc*Sf;
kA = 1./(1+P.*KD); %Retention Factor!

%  Nturn = linspace(0,Tau,t); % number of turnover


[Ncup el_time comp] = size(Xcm);   %Ccup profile (Ncup,time,species)
vmcup = Vm/Ncup;
tCM = el_time;
VCMspan = vmcup.*linspace(1,tCM,tCM); % Vspan for previous CM  

tspan = fix(ceil(Ncup*1.3)); % timespan for MP (Ncup) + SP (Ncup*Sf);x1.3 to give enough elution time
Y = zeros(Ncup, tspan,  comp); %SP Concentration g/L
X = zeros(Ncup, tspan, comp); % MP concentration

%load previous concentration profiles from classic elution to set initial boundary
Xin = zeros(Ncup,1,comp);
Yin = zeros(Ncup,1,comp);


for h = 1:length(KD)
%     X_elution(:,:,comp) = Xcm(:,:,comp); % Keep use the same format (i,t,j)
%     Y_elution(:,:,comp) =  Ycm(:,:,comp);               %KD(comp).*X_elution(:,:,comp);
    Xin(:,1,h) = Xcm(:,el_time,h);
    Yin(:,1,h) = Ycm(:,el_time,h);       %KD(comp).*Xin(:,1,comp);
    
end

Vcell = Vc/Ncup;
% Vmcup = Vm/Ncup;

% n = length(KD);

for j = 1:comp  %comp
    
%EECPC start - displace one cell at a given timestep by new SP

%     t=1; %initial condition at t = 1
    Y(1,1,j) = 0;
    X(1,1,j) = 0;
% Y(1,1,j) = Sf*Yin(1,1,j);


    for i = 2:Ncup
        
        % X(i,1,j) = kA(j)*(Xin(i-1,1,j)+P*Yin(i,1,j));
        % Y(i,1,j) = KD(j)*X(i,1,j);

        X(i,1,j) = Xin(i-1,1,j);
        Y(i,1,j) = Yin(i-1,1,j);

    end

              
     for t = 2:Ncup % MP elution end at Ncup time 
                
          for i = 2:Ncup      
           
% one cell is displaced by SP after one turnover                    
                X(i,t,j) = X(i-1,t-1,j);
                Y(i,t,j) = Y(i-1,t-1,j);
                  
                 
          end
          
         
     end
     

     %save outlet concentration 
         for h = 1:comp
            
                CoutCM (h,:) = Xcm(Ncup,:,h);
            
               for t = 1:tspan
               
                   CoutE(h,t) = Sf*Y(Ncup,t,h)+(1-Sf)*X(Ncup,t,h);
                    
               end
               % 
         end

  end

                  Cout = [CoutCM   CoutE];  % ouput effluent histories


     %Combine concentration profiles into Xtot & Ytot from CM to EECCC
          Xtot = [Xcm  X];
          Ytot = [Ycm  Y];
     
        
          
          diff = tspan; %turnover time step
          texrusion = linspace(1, diff, diff);
          Vextrusion = Vcell.*texrusion;

          V_EECPC = VCMspan(end) + Vextrusion;

          Vspan = [VCMspan   V_EECPC]; %total elution volume
            
          % V_boundary = [VCMspan(end)   VCMspan(end)+Vsweep(end)];
          V_boundary = [VCMspan(end)   Vspan(end)];

end

          
     
  

File: models/ECPC_V1.py
from numba import jit, prange
import numpy as np


@jit(nopython=True, parallel=True)
def ECPC_V1(KD, Vc, Sf, Xcm, Ycm, extrusion_steps=None):
    """
    Parameters:
    KD: Distribution coefficient ([Conc_SP]eq/[Conc_MP]eq)
    Vc: Total column volume
    Sf: Stationary factor (volume ratio of [V_SP]/[V_MP])
    Xcm: MP profiles from CM (Ncup x el_time x comp)
    Ycm: SP profiles from CM (Ncup x el_time x comp)

    Returns:
    Vspan: Total elution volume
    Cout: Elution volume vs. column outlet concentration
    Xtot: Total MP profiles (CM + ECPC)
    Ytot: Total SP profiles (CM + ECPC)
    V_boundary: CM and extrusion boundary in terms of elution volume
    """
    P = Sf / (1 - Sf)  # Phase ratio
    Vm = Vc * (1 - Sf)  # MP volume
    kA = 1 / (1 + P * KD)  # Retention factor

    Ncup, el_time, comp = Xcm.shape  # Cup profile dimensions
    vmcup = Vm / Ncup
    tCM = el_time

    # VCMspan for previous CM
    VCMspan = np.zeros(tCM)
    for i in range(tCM):
        VCMspan[i] = vmcup * (i + 1)

    if extrusion_steps is not None:
        tspan = int(Ncup + extrusion_steps)
    else:
        tspan = int(np.ceil(Ncup * (1 + Sf) * 1.3))

    Y = np.zeros((Ncup, tspan, comp))  # SP concentration
    X = np.zeros((Ncup, tspan, comp))  # MP concentration

    Vcell = Vc / Ncup

    # Process each component in parallel
    for j in prange(comp):
        # ECPC start - displace one cell at a given timestep by new SP
        Y[0, 0, j] = 0
        X[0, 0, j] = 0

        for i in range(1, Ncup):
            X[i, 0, j] = Xcm[i - 1, el_time - 1, j]
            Y[i, 0, j] = Ycm[i - 1, el_time - 1, j]

        # MP elution ends at Ncup time
        for t in range(1, Ncup):
            for i in range(1, Ncup):
                # One cell is displaced by SP after one turnover
                X[i, t, j] = X[i - 1, t - 1, j]
                Y[i, t, j] = Y[i - 1, t - 1, j]

    # Save outlet concentration
    Cout = np.zeros((comp, el_time + tspan))
    for h in range(comp):
        # CM outlet
        for t in range(el_time):
            Cout[h, t] = Xcm[Ncup - 1, t, h]
        # ECPC outlet
        for t in range(tspan):
            Cout[h, el_time + t] = Sf * Y[Ncup - 1, t, h] + (1 - Sf) * X[Ncup - 1, t, h]

    # Combine concentration profiles into Xtot & Ytot from CM to ECPC
    Xtot = np.zeros((Ncup, el_time + tspan, comp))
    Ytot = np.zeros((Ncup, el_time + tspan, comp))

    # Copy CM profiles
    for i in range(Ncup):
        for t in range(el_time):
            for j in range(comp):
                Xtot[i, t, j] = Xcm[i, t, j]
                Ytot[i, t, j] = Ycm[i, t, j]

    # Copy ECPC profiles
    for i in range(Ncup):
        for t in range(tspan):
            for j in range(comp):
                Xtot[i, el_time + t, j] = X[i, t, j]
                Ytot[i, el_time + t, j] = Y[i, t, j]

    # Combine elution volume
    diff = tspan  # Turnover time step
    Vextrusion = np.zeros(diff)
    for i in range(diff):
        Vextrusion[i] = Vcell * (i + 1) + VCMspan[-1]

    # Total elution volume
    Vspan = np.zeros(el_time + tspan)
    for i in range(el_time):
        Vspan[i] = VCMspan[i]
    for i in range(tspan):
        Vspan[el_time + i] = Vextrusion[i]

    # CM and extrusion boundary
    V_boundary = np.array([VCMspan[-1], Vspan[-1]])

    return Vspan, Cout, Xtot, Ytot, V_boundary
